Rmk88  split screen updates and color, eliminate reliance on STREAMHINT xpointer

rmkaplan opened the pull request:
As usual, a lot of files changed.  Internal cleanup was mostly aimed at eliminating all internal dependencies on the STREAMHINT xpointer from the TEXTOBJ to the TEXTSTREAM.  The xpointer is still there, against the possibility of client code still getting the TEXTOBJ object.  But that code must separately hold on to the stream, else it will get collected.

This presents a menu-based interface for splitting and unsplitting windows, without reserving the right-edge slice of the window for that purpose (and the mouse tracking and cursor switching to a little square box).  The old behavior is still there, controlled by the variable TEDIT.OP.WIDTH (was 12, near 0).

This supports color (although untested) as a character-looks property COLOR, BLACK by default.  

See TEDIT-RELEASENOTES.

rmkaplan commented:
I changed the filel loadup order to avoid the "undefined" messages in fuller.database

Also, another internal change worth mentioning is to the THISLINE vector:  Instead of in effect runcoding the character-looks changes across a line, there is now a separate field for the character looks of each character.  So THISLINE now supports random access, making it easier to do backward adjustments (eventually for kerning).   In a next version, I think the running X position should be stored as a separte field so that the displayer doesn't compute it on the fly from the width.

MattHeffron commented on library/tedit/TEDIT-FNKEYS:
```diff
@@ -88,13 +90,13 @@
 
     (if (LISTP CODE)
-        then (CONS (CHARNAME (CAR CODE)
+        then (CONS (CHARCODE.ENCODE (CAR CODE)
```
You're not passing the `NONCHARIDENTITY` to the recursive calls.

MattHeffron commented on library/tedit/TEDIT-FNKEYS:
```diff
@@ -108,21 +110,23 @@
                  (CHAR (LOGAND CODE 255))
                  (ASCIICODE (LOGAND CODE 127))
-                 CSETNAME CHARNAME ASCIINAME)
+                 CSETNAME CHARCODE.ENCODE ASCIINAME)
```
Renaming the variable here seems like a chance for confusion.

rmkaplan commented:
Updated CHARCODE.ENCODE to pass NONCHARIDENTITY in recursive calls, also reverted internal variable name to CHARNAME (previous replace was over-eager)

pamoroso commented:
On Linux Mint 22.1 Cinnamon I tested these changes up to a8f2fcc.

Window splitting works well but I still notice the mouse pointer turns to a square when hovering close to the right edge of the window. I don't mind this and it causes no issues, it's just an observation.

I get the error `TEDIT.OP.WIDTH is an unbound variable.` when evaluating `TEDIT.OP.WIDTH`.

rmkaplan commented:
It's \TEDIT.OP.WIDTH (backslash), sorry for the mistake.  

I set it to 0 to turn it off by default, I think now I have to set it to -1 to avoid it still popping up.

pamoroso commented:
After experimenting a bit with window splitting as [mentioned earlier](https://github.com/Interlisp/medley/pull/2119#issuecomment-2828075006) I executed the `Exit` title bar command of the TEdit window I was experimenting with. I experimented with the file TEDIT-RELEASENOTES.TEDIT open in TEdit divided into two panes but at the end of the session I unsplit the lower one. Yet after `Exit` TEdit quit in the state shown in this screenshot:

![tedit-split-window](https://github.com/user-attachments/assets/7ea304f3-2847-4bec-b38e-7321488f8a35)

pamoroso commented:
I confirm `\TEDIT.OP.WIDTH` works.

A clarification on my latest post: to quit TEdit I used the `Quit` command, not `Exit`.

rmkaplan commented:
I don't see that behavior.  I tried

TEDIT(TEDIT-RELEASENOTES)
select a character halfway down
click split
click unsplit (caret was still blinking in the lower window)
click quit

The screen is clear.

I also tried the same sequence, but using the window-command menu to click Close (not the Tedit menu to Quit).

The screen is still clear.

Is there anything else going on?

> On Apr 24, 2025, at 8:48 AM, Paolo Amoroso ***@***.***> wrote:
> 
> 
> pamoroso
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2828104909>
> I confirm \TEDIT.OP.WIDTH works.
> 
> A clarification on my latest post: to quit TEdit I used the Quit command, not Exit.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2828104909>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJIUUPGKADGS3NJIAXT23EBUDAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRYGEYDIOJQHE>.
> You are receiving this because you authored the thread.
>

pamoroso commented:
I did something similar in the test session and unsplit the lower pane prior to `Quit`ting. In the session I selected some text and copied some to the clipboard. I took no further actions or executed other commands.

By the way, other than the missing title bar the [leftover window](https://github.com/Interlisp/medley/pull/2119#issuecomment-2828097204) was an ordinary TEdit window with the document still intact and scrollable.

pamoroso commented:
I updated to 1ceb1a4 and tested again by opening TEDIT-RELEASENOTES.TEDIT in TEdit.

I split and unsplit the window several times into up to 4 panes. At some point, when unsplitting the panes one by one to return to one pane, I got a break window with this error:

```
INTERLISP-ERROR

In ERROR:
ARG NOT TEXTOBJ
NIL
```

The TEdit window disappeared leaving this corrupted screen:

![tedit-split-error](https://github.com/user-attachments/assets/a45f023f-70b0-490c-b13f-f1d825745c59)

The context of the error:

```
MOUSE/11(debug)BTV
   MESS1 "ARG NOT TEXTOBJ"
   MESS2 NIL
   NOBREAK NIL
ERROR
\LISPERROR
\DTEST.UFN
   PANE {WINDOW}#122,40000
\TEDIT.CURSORMOVEDFN
   FNLST \TEDIT.CURSORMOVEDFN
   WINDOW {WINDOW}#122,40000
   CHECKFORDON'TFLG NIL
DOUSERFNS
   SI::*DUMMY-FOR-CATCH* T
   SI::*CATCH-RETURN-FROM* (&)
WINDOW.MOUSE.HANDLERA0032
   \MHCOM \TEDIT.CURSORMOVEDFN
   \MHPROCESS NIL
   \MHWINDOW {WINDOW}#122,40000
   \MOUSEBUSY T
   SI::NLSETQ-VALUE NIL
   *PROCEED-CASES* (&)
   SI::*NLSETQFLAG* NIL
WINDOW.MOUSE.HANDLER
   \OLDTTY #<Process OLDMOUSE#5/121,127306>
   \MOUSEBUSY NIL
\MOUSE.PROCESS
   *FORM* (\MOUSE.PROCESS)
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* \MOUSE.PROCESS
\EVALFORM
   %#FORM# (\MOUSE.PROCESS)
   *CURRENT-PROCESS* #<Process MOUSE/121,127204>
   HELPFLAG BREAK!
   \CURRENTDISPLAYLINE 0
   \#DISPLAYLINES 40
   \LINEBUF.OFD #<IO Stream on T/174,72500>
   *READTABLE* #<ReadTable INTERLISP/174,71714>
   \PRIMTERMTABLE {TERMTABLEP}#174,64700
   \PRIMTERMSA {CHARTABLE}#167,74404
   TtyDisplayStream 
#<Output Display Stream/173,157000>
   SI::*RESETFORMS* NIL
   \INTERRUPTABLE T
   \TTYWINDOW NIL
   READBUF NIL
   \TERM.OFD #<Output Stream on T/170,120300>
   *STANDARD-OUTPUT* #<Output Stream on T/170,120300>
   *STANDARD-INPUT* #<IO Stream on T/174,72500>
\MAKE.PROCESS0
T
```

Medley head is mk88--Split-screen-updates-and-color:

```
commit 1ceb1a42c23cd05d61451f5234b76d4675b44027
Author: rmkaplan <ron.kaplan@post.harvard.edu>
Date:   Thu Apr 24 10:05:37 2025 -0700

    Make the split region really go away by default
```

Maiko head is master:

```
commit 4994982a90565ae18d540eb402047a9d6ef5713d
Author: Matt Heffron <heffron@alumni.caltech.edu>
Date:   Tue Feb 11 20:03:01 2025 -0800

    Adds the non-keypad arrow keys to be detected.
    With additional Medley calls to KEYACTION (below), they should work in TTYIN (i.e., XCL and Interlisp EXECs).
    I didn't try TEDIT.
    They are unrecognized in SEDIT. (In any case, their use there probably is pretty restricted by the nature of SEDIT.)
    
    (KEYACTION 'KEYPAD8 '(("Meta,^" 56 NOLOCKSHIFT) . IGNORE))
```

pamoroso commented:
I forgot to mention that I run Medley under VNC.

pamoroso commented:
I've been testing the past few commits up to a43a80c and in all of them experienced issues and errors when `Ctrl`-selecting text for deletion, for example `Ctrl`-middle-click to delete a word.

Sometimes the text is highlighted in reverse but nothing happens when releasing `Ctrl` and the mouse button. In such cases I `Ctrl`-middle-click the same text again and get an error. For example, in TEdit I wrote the text:

```
one

two

three

four

five
```

Then I `Ctrl`-middle-clicked each word and blank line until `five` was only highlighted in reverse with no deletion. `Ctrl`-middle-clicking again yielded this break window with the error:

```
TYPE-MISMATCH

In CL::%%NOT-NUMBER-ERROR:
{CL100/22470:Gacha10} is not a number
```

![tedit-delete](https://github.com/user-attachments/assets/512a9182-4afc-4fa9-b93c-64a3bc274967)

The context:

```
MOUSE/5(debug)BTV
CL::%%NOT-NUMBER-ERROR
\SLOWPLUS2
   LINE {L82/9640: *5-5* FED}
   X 15
   NEWSEL {SEL: unset 101/27238}
   SELOPERATION DELETE
   TSTREAM #<IO Text Stream/130,5500>
   BUTTON MIDDLE
   WORDSELFLG T
   TEXTOBJ {TEXTOBJ}#122,32600
   CHARSLOT NIL
   CLOOKS NIL
   CHNO 5
   X0 NIL
   XLIM 16
   DCH NIL
   SELCHAR NIL
   PASTRIGHT T
   THISLINE {THISLINE}#122,31764
   RIGHTHALF NIL
   CS {}#110,170512
   CHAR NIL
   CHARW {CL100/22470:Gacha10}
   CHARCL NIL
\TEDIT.SCAN.LINE
   X 30
   Y 366
   NEWSEL {SEL: unset 101/27238}
   TSTREAM #<IO Text Stream/130,5500>
   SELOPERATION DELETE
   PANE {WINDOW}#122,26664
   BUTTON MIDDLE
   CURSEL {SEL: 1-3-R  101/27204}
   REGIONTYPE NIL
   TEXTOBJ {TEXTOBJ}#122,32600
   LINE {L82/9640: *5-5* FED}
   PARAFIRSTCHNO NIL
   PARALASTCHNO NIL
   SELFN NIL
\TEDIT.XYTOSEL
   SI::*CLEANUP-FORMS* SI::RESETUNWIND
   TTYPROC #<Process TEdit/174,24000>
   TSTREAM #<IO Text Stream/130,5500>
   X 30
   Y 366
   DS #<Output Display Stream/130,5700>
   OLDX 30
   OLDY 366
   PREG (0 0 482 380)
   TEXTOBJ {TEXTOBJ}#122,32600
   CURSEL {SEL: 1-3-R  101/27204}
   NEWSEL {SEL: unset 101/27238}
   CUROPERATION DELETE
   NEWOPERATION DELETE
   PENDINGDEL NIL
   READONLY NIL
SI::*UNWIND-PROTECT*
   PANE {WINDOW}#122,26664
   LISPXHIST NIL
   SI::*RESETFORMS* ((& {CURSOR}#157,141334))
   RESETSTATE NIL
MODERN-ORIG-\TEDIT.BUTTONEVENTFN
   WINDOW {WINDOW}#122,26664
   ORIGFUNCTION MODERN-ORIG-\TEDIT.BUTTONEVENTFN
   ANYWHERE NIL
   TITLEPROPORTION NIL
   CORNERREGION (756 358 490 400)
   TOPMARGIN 12
   CORNER NIL
   ATTACHEDREGION NIL
MODERNWINDOW.BUTTONEVENTFN
   WINDOW {WINDOW}#122,26664
   ORIGFUNCTION MODERN-ORIG-\TEDIT.BUTTONEVENTFN
   ANYWHERE NIL
   TITLEPROPORTION NIL
   CORNERREGION (756 358 490 400)
   TOPMARGIN "* Tedit Window"
MODERNWINDOW.BUTTONEVENTFN
   W {WINDOW}#122,26664
   STREAM NIL
\TEDIT.BUTTONEVENTFN
   W {WINDOW}#122,26664
   STREAM NIL
\TEDIT.BUTTONEVENTFN
   SI::*DUMMY-FOR-CATCH* T
   SI::*CATCH-RETURN-FROM* (&)
\PROTECTED.APPLYA0001
   FN \TEDIT.BUTTONEVENTFN
   WINDOW {WINDOW}#122,26664
   SI::NLSETQ-VALUE NIL
   *PROCEED-CASES* (&)
   SI::*NLSETQFLAG* NIL
\PROTECTED.APPLY
   \MHCOM \TEDIT.BUTTONEVENTFN
   \MHPROCESS #<Process TEdit/174,24000>
   \MHWINDOW {WINDOW}#122,26664
   \MOUSEBUSY T
WINDOW.MOUSE.HANDLER
   \OLDTTY #<Process TEdit/174,24000>
   \MOUSEBUSY NIL
\MOUSE.PROCESS
   *FORM* (\MOUSE.PROCESS)
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* \MOUSE.PROCESS
\EVALFORM
   %#FORM# (\MOUSE.PROCESS)
   *CURRENT-PROCESS* #<Process OLDMOUSE/174,24410>
   HELPFLAG BREAK!
   \CURRENTDISPLAYLINE 0
   \#DISPLAYLINES 40
   \LINEBUF.OFD #<IO Stream on T/174,72500>
   *READTABLE* #<ReadTable INTERLISP/174,71714>
   \PRIMTERMTABLE {TERMTABLEP}#174,64740
   \PRIMTERMSA {CHARTABLE}#174,65000
   TtyDisplayStream 
#<Output Display Stream/146,12300>
   SI::*RESETFORMS* NIL
   \INTERRUPTABLE T
   \TTYWINDOW NIL
   READBUF NIL
   \TERM.OFD #<Output Stream on T/170,116300>
   *STANDARD-OUTPUT* #<Output Stream on T/170,116300>
   *STANDARD-INPUT* #<IO Stream on T/174,72500>
\MAKE.PROCESS0
T
```

Aside from that I no longer get crashes when splitting and unsplitting. In addition the mouse pointer no longer turns into a square when hovering close to the right edge of the window.

rmkaplan commented:
I pushed some a few changes having to do with the menu bar behavior for tabs.  I also fixed up some mentions of XCCS on the way to MCCS.

Note that there is currently a pdf hardcopy problem with this PR:  The default color is BLACK, and that is one of the ways of specifying colors according to the documentation (man color).  But postscript (\DSPCOLOR.PSC) only accepts a number between 0 and 1, which is not a legitimate way of specifying colors.  So it fails.

Immediate fix is to MOVD(NILL \DSPCOLOR.PSC).  But probably the argument checking should be installed in the generic DSPCOLOR, bringing in some of the predicates from LLCOLOR.  And postscript eventually should be changed to approximate more of what can be specified.

rmkaplan commented:
@pamoroso I have not been able to replicate the faillure with your one-two... example.

The closest I have come is to type a single line "one" with no EOL into an empty window, then ctrl-middle-select.  The screen doesn't get updated (the letters o n e are still there), and they disappear with redisplay or with typing in.  So I think I missed a case where the line should be updated, but it doesn't cause an error.

Is there any more specific recipe to get the break?

MattHeffron commented:
@rmkaplan See my PR #2127 for my attempt to support correct color specifications.

rmkaplan commented:
@pamoroso I believe I have fixed the screen update for an edit (delete) to the top line.  I still don't get a break.

pamoroso commented:
I updated Medley to 54f1768 and built it with `loadup-all.sh -apps` but it broke with the error `Can't find package CLOS to look up symbol LOAD-CLOS`:

![clos-error](https://github.com/user-attachments/assets/809ddc59-001e-4746-86b9-df55db5aec06)

```
>>>>> START loadup-full-from-lisp
"/home/paolo/medley/medley/../maiko/linux.x86_64/lde" "/tmp/loadups-17757/lisp.sysout" -id "loadup_1" -title "Medley Interlisp :: loadup_1" -g 1024x768 -sc 1024x768 -noscroll
MEDLEYDIR: "/home/paolo/medley/medley"
LOGINDIR: "/tmp/loadups-17757/logindir"
GREET FILE: "/home/paolo/medley/medley/greetfiles/NOGREET"
REM.CM FILE: "/tmp/loadups-17757/loadup-full-from-lisp.cm"
VMEM FILE: "/tmp/loadups-17757/logindir/vmem/lisp_loadup_1.virtualmem"
+++++ SUCCESS +++++
..... files created .....
-rw-rw-r-- 1 paolo paolo 10166 Apr 27 11:06 /tmp/loadups-17757/full.dribble
-rw-rw-r-- 1 paolo paolo 10846208 Apr 27 11:06 /tmp/loadups-17757/full.sysout
<<<<< END loadup-full-from-lisp

>>>>> START loadup-apps-from-full
"/home/paolo/medley/medley/../maiko/linux.x86_64/lde" "/tmp/loadups-17757/full.sysout" -id "loadup_1" -title "Medley Interlisp :: loadup_1" -g 1024x768 -sc 1024x768 -noscroll
MEDLEYDIR: "/home/paolo/medley/medley"
LOGINDIR: "/tmp/loadups-17757/logindir"
GREET FILE: "/home/paolo/medley/medley/greetfiles/NOGREET"
REM.CM FILE: "/tmp/loadups-17757/loadup-apps-from-full.cm"
VMEM FILE: "/tmp/loadups-17757/logindir/vmem/lisp_loadup_1.virtualmem"
```

Here are the dribble files: [clos-error-dribble.zip](https://github.com/user-attachments/files/19928802/clos-error-dribble.zip)

I haven't changed anything in my Linux Mint 22.1 Cinnamon setup and Maiko is on master/4994982.

rmkaplan commented:
I believe I have fixed the break reported by @pamoroso .  It had to do with clicking on an empty line at the end of the file, not related to deletion or middle-button.

pamoroso commented:
I updated to 3136cd8 but the build still fails with the [same error](https://github.com/Interlisp/medley/pull/2119#issuecomment-2833345255).

rmkaplan commented:
Is this only in the apps loadup?  I think only notecards loads the tedit process killer, if that's where the problem lies.

pamoroso commented:
Yes I get the error only in the apps loadup, building with `loadup-all.sh` completes successfully.

I tested TEdit and no longer get breaks when clicking on an empty line at the end of the file. `Ctrl`-middle-selecting works as expected too.

rmkaplan commented:
@pamoroso I restored the function \TEDIT.ACTIVE.WINDOWP.  lispusers/TEDIT-PROCESS-KILLER (which is not in the MS database) was its only client, and the function was out of date.  Try the apps loadup again

nbriggs commented:
I find `grep -ri "foo" *` (from the medley directory) to be more reliable than MS at finding uses, and fast enough (10s) even though it's searching a lot of data - e.g.

```code
briggs@macrobiotic medley % grep -ri '\\TEDIT.ACTIVE.WINDOWP'  *
Binary file internal/venuesysouts/full.venuesysout matches
Binary file library/tedit/TEDIT-WINDOW.LCOM matches
library/tedit/TEDIT-WINDOW:        (FNS \TEDIT.CURSORMOVEDFN \TEDIT.CURSOROUTFN \TEDIT.ACTIVE.WINDOWP \TEDIT.EXPANDFN 
library/tedit/TEDIT-WINDOW:(\TEDIT.ACTIVE.WINDOWP
library/tedit/TEDIT-WINDOW:\TEDIT.ACTIVE.WINDOWP 51177 . 52228) (\TEDIT.EXPANDFN 52230 . 52793) (\TEDIT.MAINW 52795 . 54075) (
Binary file lispusers/TEDIT-PROCESS-KILLER.LCOM matches
lispusers/tedit-process-killer:                                               (PROCESSP :IN \TEDIT.ACTIVE.WINDOWP))
lispusers/tedit-process-killer:[XCL:REINSTALL-ADVICE '(PROCESSP :IN \TEDIT.ACTIVE.WINDOWP)
lispusers/tedit-process-killer:       '((:LAST (RETURN (OR !VALUE (WITHOUT-TEDIT-PROCESS (STKARG 'W '\TEDIT.ACTIVE.WINDOWP]
lispusers/tedit-process-killer:       (PROCESSP :IN \TEDIT.ACTIVE.WINDOWP))
Binary file loadups/whereis.hash matches
Binary file loadups/full.sysout.~2~ matches
Binary file loadups/full.sysout.~1~ matches
Binary file loadups/full.sysout.~3~ matches
Binary file loadups/apps.sysout matches
Binary file loadups/full.sysout matches
Binary file loadups/build/whereis.hash matches
Binary file loadups/build/full.sysout.~2~ matches
Binary file loadups/build/full.sysout.~1~ matches
Binary file loadups/build/full.sysout.~3~ matches
Binary file loadups/build/apps.sysout matches
Binary file loadups/build/full.sysout matches
Binary file loadups/full.sysout.~4~ matches
obsolete/library/new/TEDIT:                                           (\TEDIT.ACTIVE.WINDOWP TEDIT.DEFAULT.WINDOW))
Binary file obsolete/library/new/TEDIT.LCOM matches
Binary file obsolete/library/new/TEDITWINDOW.LCOM matches
```

pamoroso commented:
Updating to 2ecea94 builds Medley successfully. But after a series of `Ctrl`-middle-select tests as described above, when quitting TEdit with `Quit` and confirming with one or two left-clicks TEdit closes and this break window shows up:

```
INTERLISP-ERROR

ARG NOT TEXTOBJ
NIL
```

![tedit-quit-error](https://github.com/user-attachments/assets/5b99cf32-b243-4b1c-bc56-01d7a2dc79c6)

The backtrace:

```
MOUSE/5(debug)BTV
   MESS1 "ARG NOT TEXTOBJ"
   MESS2 NIL
   NOBREAK NIL
ERROR
\LISPERROR
\DTEST.UFN
   PANE {WINDOW}#122,23264
\TEDIT.CURSORMOVEDFN
   FNLST \TEDIT.CURSORMOVEDFN
   WINDOW {WINDOW}#122,23264
   CHECKFORDON'TFLG NIL
DOUSERFNS
   SI::*DUMMY-FOR-CATCH* T
   SI::*CATCH-RETURN-FROM* (&)
WINDOW.MOUSE.HANDLERA0032
   \MHCOM \TEDIT.CURSORMOVEDFN
   \MHPROCESS NIL
   \MHWINDOW {WINDOW}#122,23264
   \MOUSEBUSY T
   SI::NLSETQ-VALUE NIL
   *PROCEED-CASES* (&)
   SI::*NLSETQFLAG* NIL
WINDOW.MOUSE.HANDLER
   \OLDTTY #<Process TEdit/122,16614>
   \MOUSEBUSY NIL
\MOUSE.PROCESS
   *FORM* (\MOUSE.PROCESS)
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* \MOUSE.PROCESS
\EVALFORM
   %#FORM# (\MOUSE.PROCESS)
   *CURRENT-PROCESS* #<Process OLDMOUSE/174,24410>
   HELPFLAG BREAK!
   \CURRENTDISPLAYLINE 0
   \#DISPLAYLINES 40
   \LINEBUF.OFD #<IO Stream on T/174,72500>
   *READTABLE* #<ReadTable INTERLISP/174,71714>
   \PRIMTERMTABLE {TERMTABLEP}#174,64740
   \PRIMTERMSA {CHARTABLE}#174,65000
   TtyDisplayStream 
#<Output Display Stream/130,12100>
   SI::*RESETFORMS* NIL
   \INTERRUPTABLE T
   \TTYWINDOW NIL
   READBUF NIL
   \TERM.OFD #<Output Stream on T/170,116300>
   *STANDARD-OUTPUT* #<Output Stream on T/170,116300>
   *STANDARD-INPUT* #<IO Stream on T/174,72500>
\MAKE.PROCESS0
T
```

pamoroso commented:
The [TEdit issue](https://github.com/Interlisp/medley/pull/2119#issuecomment-2834081483) is reproducible more easily:

1. open TEdit
2. type a space
3. execute the `Quit` command
4. left-click once or twice to confirm quitting (until TEdit closes and the break window shows up)

rmkaplan commented:
@pamoroso  Try this one.

pamoroso commented:
I confirm af7934e resolves the TEdit quitting issue.

pamoroso commented:
I tested up to dee919d and I have no issues to report.

rmkaplan commented:
I removed the draft status, I haven't encountered or heard of any new issues.

pamoroso commented:
On Linux Mint Cinnamon 22.1 I updated to 3520893, opened TEDIT-RELEASENOTES.TEDIT, and split the TEdit window into 7 panes. Then I unsplit the panes in random order until only the full main pane remained in the window. When I `Quit` TEdit I ran into the same issue [reported earlier](https://github.com/Interlisp/medley/pull/2119#issuecomment-2828097204), i.e. leftover windows:

![tedit-split-issue](https://github.com/user-attachments/assets/ed67454f-ac60-4974-b4d9-d88100d6e858)

rmkaplan commented:
@pamoroso , I still haven't been able to replicate this.  I wrote a test jig that opens the file, splits it at the same 7 places in ascending order on every run, but then unsplits the panes in a different random order each time.  At the end it executes the function that the Quit menu item uses to say that the editing session is over, and then the document/window cleans itself up and disappears.

I've run this maybe 20 times, and it hasn't yet left any orphan panes on the screen.

Is there something else that I may be missing?

(Meanwhile, I tested color against Herb's upcoming HTML stream, and that works)

rmkaplan commented:
@pamoroso , I am now seeing some examples, will work on it.

rmkaplan commented:
I believe I have figured out and eliminated the behavior of split windows showing up on the screen when the main window has been closed.

This was a hard one because it was nondeterministic, sometimes no window resurfaced, sometimes more than one.  At least one cause (I hope there aren't others) was the fact that the Tedit caret flashing was not being disabled at the right time as the window was being closed, and the caret caused the window to pop back up as an orphan.  With this update I have so far not been able to provoke the bad behavior.

pamoroso commented:
On Linux Mint Cinnamon 22.1 I updated to 3d67bb1, opened the TEdit release notes in TEdit, executed the split command and got this error:

```
ATTEMTP-TO-RPLAC-NIL
In \RPLACD.UFN:
Attempt to rplac NIL with ({WINDOW}#121,120400 {L81/39294: *72-116* FE})
```

![tedit-split-error](https://github.com/user-attachments/assets/96723eee-c579-4de9-a782-050849c2c409)

The backtrace:

```
TEDIT.DEFAULT.MENUFN/6(debug)BTV
\RPLACD.UFN
   SEL {SEL: 92-1-R _ 101/26418}
   TSTREAM #<IO Text Stream/146,10300>
   ONLYPANE {WINDOW}#121,120400
   TEXTOBJ {TEXTOBJ}#122,11600
   FIRSTCHNO 92
   LASTCHNO 92
   PANE {WINDOW}#121,120400
   PSTARTLINE {L81/39294: *72-116* FE}
   PENDLINE {L81/39294: *72-116* FE}
   X0 115
   XLIM 119
   L1 ({L81/39294: *72-116* FE})
   LN ({L81/39294: *72-116* FE})
\TEDIT.FIXSEL
   PANE {WINDOW}#121,120400
   TSTREAM #<IO Text Stream/146,10300>
   PROPS (NOTITLE T PROMPTWINDOW DON'T TITLEMENUFN 
NILL)
   AFTERPANE {WINDOW}#122,21000
   LCHAR1 72
   TEXTOBJ {TEXTOBJ}#122,11600
   MENUPROP NIL
   SEL {SEL: 92-1-R _ 101/26418}
   LASTVISIBLE NIL
\TEDIT.WINDOW.SETUP
   OLDPANE {WINDOW}#122,21000
   Y 768
   TSTREAM #<IO Text Stream/146,10300>
   TEXTOBJ {TEXTOBJ}#122,11600
   PREG (731 23 516 933)
   SEL {SEL: 92-1-R _ 101/26418}
   NEXTPANE NIL
   ATTACHEDWINDOWS ({WINDOW}#121,120664)
   NEWPANE {WINDOW}#121,120400
   PROPS (NOTITLE T PROMPTWINDOW DON'T TITLEMENUFN 
NILL)
   NEXTCHAR1 72
\TEDIT.SPLITW
   SI::*CLEANUP-FORMS* SI::RESETUNWIND
SI::*UNWIND-PROTECT*
   SI::*DUMMY-FOR-CATCH* T
   SI::*CATCH-RETURN-FROM* (&)
   LISPXHIST NIL
   SI::*RESETFORMS* NIL
   RESETSTATE NIL
TEDIT.DEFAULT.MENUFNA0001
   PANE {WINDOW}#122,21000
   TSTREAM #<IO Text Stream/146,10300>
   TEXTOBJ {TEXTOBJ}#122,11600
   WMENU NIL
   ITEM Split% Window
   THISMENU {MENU}#145,7540
   SI::NLSETQ-VALUE NIL
   *PROCEED-CASES* (&)
   SI::*NLSETQFLAG* NIL
TEDIT.DEFAULT.MENUFN
   *FORM* (TEDIT.DEFAULT.MENUFN (QUOTE {WINDOW}#122,
21000))
   *ARGVAL* NIL
   *TAIL* NIL
   *FN* TEDIT.DEFAULT.MENUFN
\EVALFORM
   %#FORM# (TEDIT.DEFAULT.MENUFN (QUOTE {WINDOW}#122,
21000))
   *CURRENT-PROCESS* 
#<Process TEDIT.DEFAULT.MENUFN/121,107614>
   HELPFLAG BREAK!
   \CURRENTDISPLAYLINE 0
   \#DISPLAYLINES 40
   \LINEBUF.OFD #<IO Stream on T/174,73500>
   *READTABLE* #<ReadTable INTERLISP/174,72714>
   \PRIMTERMTABLE {TERMTABLEP}#174,65740
   \PRIMTERMSA {CHARTABLE}#174,66000
   TtyDisplayStream 
#<Output Display Stream/127,106500>
   SI::*RESETFORMS* NIL
   \INTERRUPTABLE T
   \TTYWINDOW NIL
   READBUF NIL
   \TERM.OFD #<Output Stream on T/170,114300>
   *STANDARD-OUTPUT* #<Output Stream on T/170,114300>
   *STANDARD-INPUT* #<IO Stream on T/174,73500>
\MAKE.PROCESS0
T
```

rmkaplan commented:
Oops, I uploaded the wrong version

masinter commented:
this PR seems like it's gotten really big and difficult to review. I'd like to make sure the USER.CM location is at least as good as the one in @pmcjones 's bravo converter in finding an appropriate user.cm for .bravo files in the CHM PARC Archives files. I will raise reviewing that in a separate issue.

rmkaplan commented:
Last commit addesses one other complexity with split windows.  Tedit allows you to provide an active editing window to be reused for a new file.  For example, (TEDIT 'GITFNS), then (TEDIT 'EXAMINEDEFS [gitfns window]).  It takes down the old process, starts up a new one, etc.

But if the existing editing window is split into a bunch of subwindows, the only sensible thing to do is to restore the window to its unsplit state, delete the process, etc. before bringing up the new file.

pamoroso commented:
I updated to a999a10 on Linux Mint 22.1 Cinnamon and I don't notice anything unusual.

masinter commented:
I've verified that if I do a loadup in a clean clone of this repo using the command

```
time ./scripts/loadup-all.sh -apps && time ./scripts/loadup-db.sh && time ./scripts/do_hcfiles.sh
```
It will die in a URAID call complaning trying to decrement a reference count less than 0.   
THe error doesn't occur when I load up with master, but with  1e758ad63d (rmk88).

With a different configuration (with merge-open-prs.sh) I got a different error which was also consistent with gc freeing something that was actually in use.

Anyway, this PR should be marked as DRAFT until the GC problem is found and fixed.

masinter commented:
maybe I updated the branch wrong? I did a rebase rather than a merge commit to update the branch to the current master.

rmkaplan commented:
OK, thanks.

I'm doing stress testing on different phases (loading, lineformatting, line display, hardcopy) to see if I can narrow it down. 

> On May 30, 2025, at 6:04 PM, Larry Masinter ***@***.***> wrote:
> 
> 
> masinter
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2923864502>
> I'm still getting a RAID call running HCFILES in a clean branch; the error messages you'll get might vary because the GC delays doing anything going wrong until more virtual storage was allocated.
> Now I get a segmentation fault.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2923864502>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJKE7XVB77FISVECGND3BD5ZNAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRTHA3DINJQGI>.
> You are receiving this because you were mentioned.
>

nbriggs commented:
How do I run a test that should fail?

rmkaplan commented:
Do a loadup in rmk88, load library/tedit/TEDIT-STRESS.LCOM, cd lispusers/, then (STRESSDISPLAY) or (STRESSINSERT) it should fail pretty soon.

If you first (SETQ CHECKARRAYS 'RECLAIM) it should keep going.

This is doing the reclaims on every check, not just the one.

> On Jun 3, 2025, at 4:37 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937730432>
> How do I run a test that should fail?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937730432>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJKFZYGQAEWLPENKCIL3BYWTTAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZXG4ZTANBTGI>.
> You are receiving this because you were mentioned.
>

nbriggs commented:
So far, 10 passes of STRESSDISPLAY with no problem.  How big is your medley display?  It puts the TEDIT display up a little bit off-screen for me. My display is 1100x700, non-scrollable.

rmkaplan commented:
1920 x 1080.  The display isn't particularly interesting, although you can see things flash by.  You can edit STRESSDISPLAY to move the window (you may have to load TEDIT-EXPORTS.ALL to run/compile that).

I just tried STRESSINSERT in what I think is the same code as in the repo. I get a failure when it gets to PS-SKETCH-PATCH on the first rep.



> On Jun 3, 2025, at 5:06 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937792986>
> So far, 10 passes of STRESSDISPLAY with no problem. How big is your medley display? It puts the TEDIT display up a little bit off-screen for me. My display is 1100x700, non-scrollable.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937792986>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJP4TOYRGPMVCT6XNSL3BY2BTAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZXG44TEOJYGY>.
> You are receiving this because you were mentioned.
>

rmkaplan commented:
Hmm, I vaguely recall seeing a lastvalid failure on one of my runs, but I usually see a failure inside one of the calls to \PARSEARRAYSPACE under CHECKARRAYS.  CHECKARRAYS has a tag argument so you can see which call it was, if you end up in URAID.  

> On Jun 3, 2025, at 5:18 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937825982>
> STRESSINSERT died with NIL is not a non-complex-number for (IGREATERP 76 NIL) called from \TEDIT.LASTVALIDLINE with FIRSTCHANGEDCHNO 76, LASTCHANGEDCHNO 78. It was trying to insert "aaa" into TRAJECTORY-FOLLOWER.TEDIT
> 
> image.png (view on web) <https://github.com/user-attachments/assets/7de70201-5551-4e28-87eb-646b26b85caa>
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2937825982>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJPFTKOGOVSELEUHJL33BY3N5AVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZXHAZDKOJYGI>.
> You are receiving this because you were mentioned.
>

nbriggs commented:
I tried running with the current maiko master lde/ldex - started a fresh full.sysout, I have very little in my init, just loaded background-yield.  In the XCL exec, `(IL:LOAD "TEDIT-STRESS.LCOM")` then `CONN`, then `CONN projects/medley/lispusers`, then `(IL:STRESSDISPLAY)`.  It loaded DOC-OBJECTS.LCOM then DATEFORMAT-EDITOR.LCOM, and broke under IL:GDATE.  Processing BITMAP-GALLERY.TEDIT.

If I open an Interlisp Exec window and do it from there -- no problem.  If I interrupt it and go back to the XCL exec and try it again, it fails in the same way.  Something is sensitive to where it's started from that probably shouldn't be!


![image](https://github.com/user-attachments/assets/ae598aa2-6c72-4dde-8d3d-49ca95f010a1)

rmkaplan commented:
Probably the DOCOBJ image object doesn't manage its readtable or package,  unrelated to the array problem.

> On Jun 3, 2025, at 8:27 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2938270757>
> I tried running with the current maiko master lde/ldex - started a fresh full.sysout, I have very little in my init, just loaded background-yield. In the XCL exec, (IL:LOAD "TEDIT-STRESS.LCOM") then CONN, then CONN projects/medley/lispusers, then (IL:STRESSDISPLAY). It loaded DOC-OBJECTS.LCOM then DATEFORMAT-EDITOR.LCOM, and broke under IL:GDATE. Processing BITMAP-GALLERY.TEDIT.
> 
> If I open an Interlisp Exec window and do it from there -- no problem. If I interrupt it and go back to the XCL exec and try it again, it fails in the same way. Something is sensitive to where it's started from that probably shouldn't be!
> 
> image.png (view on web) <https://github.com/user-attachments/assets/ae598aa2-6c72-4dde-8d3d-49ca95f010a1>
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2938270757>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJJL6JZAJZGQEXABGNT3BZRTDAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZYGI3TANZVG4>.
> You are receiving this because you were mentioned.
>

rmkaplan commented:
I tried it without my init and STRESSINSERT didn't fail after 28 reps.  With my initial it failed in the 4th rep.

I'll try to dissect my init.


> On Jun 3, 2025, at 8:27 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2938270757>
> I tried running with the current maiko master lde/ldex - started a fresh full.sysout, I have very little in my init, just loaded background-yield. In the XCL exec, (IL:LOAD "TEDIT-STRESS.LCOM") then CONN, then CONN projects/medley/lispusers, then (IL:STRESSDISPLAY). It loaded DOC-OBJECTS.LCOM then DATEFORMAT-EDITOR.LCOM, and broke under IL:GDATE. Processing BITMAP-GALLERY.TEDIT.
> 
> If I open an Interlisp Exec window and do it from there -- no problem. If I interrupt it and go back to the XCL exec and try it again, it fails in the same way. Something is sensitive to where it's started from that probably shouldn't be!
> 
> image.png (view on web) <https://github.com/user-attachments/assets/ae598aa2-6c72-4dde-8d3d-49ca95f010a1>
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2938270757>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJJL6JZAJZGQEXABGNT3BZRTDAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZYGI3TANZVG4>.
> You are receiving this because you were mentioned.
>

rmkaplan commented:
One other thing:  I'm also running with ARRAYBLOCKCHECKING set to T, in addition to all the explicit calls to \PARSEARRAYSPACE.

nbriggs commented:
OK - ARRAYBLOCKCHECKING set to T and it fails for me - but in the Lisp implementation of \PARSEARRAYSPACE, not the Maiko array space checking in gcfinal.c -- I wonder if the code matches.

nbriggs commented:
I think the code is the same.  I don't think that anything that happens during the Lisp array space check could cause the Maiko code to [de]allocate and modify the array space the Lisp code is looking at.

rmkaplan commented:
So, when it SAYS that something is wrong, you think that something actually IS wrong?

> On Jun 5, 2025, at 10:20 AM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2945357911>
> I think the code is the same. I don't think that anything that happens during the Lisp array space check could cause the Maiko code to [de]allocate and modify the array space the Lisp code is looking at.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2945357911>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJIZOIIJ33IKP4TY6K33CB37FAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNBVGM2TOOJRGE>.
> You are receiving this because you were mentioned.
>

rmkaplan commented:
I'll save RANDSTATE on a file before each random run.  

> On Jun 5, 2025, at 7:22 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2947840605>
> I don't know if something is really wrong. I've got one here with mismatched header/trailer lengths, it's a block on the free-list. It looks wrong -- the header says the length is 98, the trailer says 228.
> 
> (arrayblock)  (arlen = 98, inuse = 0, gctype = 0, password = 5461, fwd = 4682046, bkwd = 4681358)
> Near the beginning of this free block is what looks like the trailer of a another block that appears to be of length 228, and then the header of a block of length 90. I guess that's possible if blocks were merged.
> 
If it doesn't clear blocks when they are collected and merged, that seems likely.

> It might be easier to reproduce if what it did wasn't totally random. Perhaps print the result of (RANDSET T)
> once before it starts calling (RAND ...) so that we can go back and reset the random state the next time and in theory get exactly the same sequence of numbers for the inserts.
> 
> I think I need to rewrite the C code that dumps array blocks so that I can direcly read the forward/backward pointers (right now it labels the locations with the native address, but the forward/backward pointers in the data are Lisp pointers (relative to the Lisp world memory base).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2947840605>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJOO5VJBDDL2E6QQRNL3CD3MRAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNBXHA2DANRQGU>.
> You are receiving this because you were mentioned.
>

rmkaplan commented:
On each run the  RAND state is saved (by SAVERANDSTATE) in a file RANDSTATE on the connected directory.  To do another run with the previous randstate, set USELASTRANDSTATE to T.

rmkaplan commented:
So the check on the internal block must be from the call to \CHECKARRAYBLOCK that comes because ARRAYBLOCKCHECKING is T and not from the calls under \PARSEARRAYSPACE.  Because we don't get the failure from just the \PARSEARRAYSPACE.

Both the outer and inner blocks are free, but does the freelist link through the inner one?

> On Jun 6, 2025, at 2:14 PM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2119)
>  <https://github.com/Interlisp/medley/pull/2119#issuecomment-2950811443>
> With some more helpful output in the maiko "printarrayblock()" code -- what I'm seeing looks like this
> 
> free ***@***.*** header length 447 cells =>next ***@***.***, ***@***.*** says length 447 cells
> free ***@***.*** header length 120 cells =>next ***@***.***, prev ***@***.***
>           ***@***.*** says length 447 cells
> used ***@***.*** header length 4052 cell =>next ***@***.***, prev ***@***.***
> so there's a free block at the end of another free block such that they share a trailer whose length matches the outer free block, and I'm not sure which block has a pointer to the interior free block - but someone did because it got checked and the missing/mismatched trailer caused the failure.
> 
> This says to me that there's a bug somewhere in the array block splitting/merging code in maiko. Sigh.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2119#issuecomment-2950811443>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJODXC4R44UEZAVNDVT3CIADFAVCNFSM6AAAAAB3W6HAVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNJQHAYTCNBUGM>.
> You are receiving this because you were mentioned.
>

nbriggs commented:
The stack looked like this, this last time:
```
*Error* URAID Called:
Enter the URaid
"ARRAYBLOCK Header and Trailer length don't match"

< l u
  0 :    0x13fd6 : IL:RAID
  1 :    0x13fc0 : IL:\MP.ERROR
  2 :    0x13f9a : IL:\CHECKARRAYBLOCK
  3 :    0x13f76 : IL:\CHECKARRAYBLOCK
  4 :    0x13f5a : IL:\PARSEARRAYSPACE1
  5 :    0x13f44 : IL:\PARSEARRAYSPACE
  6 :    0x13f2e : IL:CHECKARRAYS
  7 :    0x13eec : IL:STRESSDISPLAY
  8 :    0x13ec6 : IL:\EVALFORM
  9 :    0x13ea8 : IL:FAULTEVAL
 10 :    0x13e8a : IL:\EVALFORM
 11 :    0x13e76 : IL:EVAL
 12 :    0x13e52 : IL:EVAL-INPUT
 13 :    0x13de0 : IL:DO-EVENT
 14 :    0x13daa : XCL:EXECA0001A0002
 15 :    0x13d72 : XCL:EXECA0001
 16 :    0x13ccc : IL:\DO.PROGV
 17 :    0x13c4e : XCL:EXEC
 18 :    0x13c22 : IL:\EVALFORM
 19 :    0x13c08 : CL:PROGN
 20 :    0x13be8 : IL:\EVALFORM
 21 :    0x13b9a : IL:\MAKE.PROCESS0
 22 :    0x11802 : CL:T

< f 1
IVAR -------
  13fb6 : 0x   e  0x  1b  IL:CODE  27
  13fb8 : 0x  7f  0x9db8  CL:STRING  "ARRAYBLOCK Header and Trailer length don't match"
  13fba : 0x  45  0xbf00  IL:ARG1  {unknown}0x45bf00
  13fbc : 0x   0  0x  4c  IL:ARG2  CL:T

## STACK BF at 0x13fbe ##
[cnt=0 ]
ivar : 0x3fb6
>> Bf's ivar says 0x13fb6 vs. IVar says 0x13fce
Fname is IL:\MP.ERROR
## STACK FX at 0x13fc0 ##
[cnt = 0 ]
 #alink           0x3fa4 
 fnhead   0x2e53c4 
 nextblock        0x3fce 
 pc               0x5d 
 nametbl  0x16bf00 
 #blink           0x0 
 #clink           0x0 
  13fca : 0x   0  0x   0  
  13fcc : 0x   0  0x  4c  

```

nbriggs commented:
I'm looking at it, and writing code (in C) to help me look at the structure of array space.  It's not done yet, but I'm using it on an original Venue lisp.sysout.

nbriggs commented:
Tried with CHECKARRAY set NIL and ARRAYBLOCKCHECKING set T -- it took a lot longer (39 rounds of STRESSDISPLAY, then 222 rounds of STRESSINSERT) - to get a header/trailer length mismatch, and it's (again) coming from the Lisp side `\CHECKARRAYBLOCK`.  The backtrace shows:
```
< l u
  0 :    0x14334 : IL:RAID
  1 :    0x1431e : IL:\MP.ERROR
  2 :    0x142f8 : IL:\CHECKARRAYBLOCK
  3 :    0x142d4 : IL:\CHECKARRAYBLOCK
  4 :    0x142aa : IL:\ALLOCBLOCK.OLD
  5 :    0x14286 : IL:\MAIKO.ALLOCBLOCK
  6 :    0x14262 : IL:BITMAPCREATE
  7 :    0x14238 : IL:\READBINARYBITMAP
  8 :    0x14220 : IL:BMOBJ.GETFN3
  9 :    0x14200 : IL:READIMAGEOBJA0001
 10 :    0x141c8 : IL:READIMAGEOBJ
 11 :    0x141a0 : IL:\TEDIT.GET.OBJECT
 12 :    0x14154 : IL:\TEDIT.GET.PIECES3
 13 :    0x13fde : IL:\TEDIT.GET.FORMATTED.FILE
 14 :    0x13fa2 : SI:*UNWIND-PROTECT*
 15 :    0x13f86 : IL:\TEDIT.OPENTEXTSTREAM.PIECES
 16 :    0x13f50 : SI:*UNWIND-PROTECT*
 17 :    0x13f34 : IL:OPENTEXTSTREAM
 18 :    0x13eee : IL:STRESSINSERT
```
`\MAIKO.ALLOCBLOCK` is the implementation that's being run for the `\ALLOCBLOCK` call at the end of `BITMAPCREATE`.

A URaid "B" block dump shows (around the header)
```
     0x1109914d0 (0x  47aa68)        0
     0x1109914d4 (0x  47aa6a)        0
     0x1109914d8 (0x  47aa6c)        0
     0x1109914dc (0x  47aa6e) aaa802f6 <- array header
     0x1109914e0 (0x  47aa70) aaa90083
     0x1109914e4 (0x  47aa72)    60006
     0x1109914e8 (0x  47aa74)    60006
     0x1109914ec (0x  47aa76)    60006
```
and around the trailer
```
     0x110992090 (0x  47b048) 406403b2
     0x110992094 (0x  47b04a) 13640103
     0x110992098 (0x  47b04c) 9c640101
     0x11099209c (0x  47b04e) 6700001b
     0x1109920a0 (0x  47b050) cff092b1
     0x1109920a4 (0x  47b052)   fe0900
     0x1109920a8 (0x  47b054)   1525b2
     0x1109920ac (0x  47b056) 2a406700
     0x1109920b0 (0x  47b058)   1bcf0a <- array trailer
     0x1109920b4 (0x  47b05a)     10d3
     0x1109920b8 (0x  47b05c) b50382b1
     0x1109920bc (0x  47b05e)   e64168
     0x1109920c0 (0x  47b060) f0b40382
     0x1109920c4 (0x  47b062) b100dd40
     0x1109920c8 (0x  47b064) 6f005ebb
     0x1109920cc (0x  47b066) e80a0000
```
Clearly this item tagged as array trailer is not actually an array trailer - it's missing the 0xaaa8 bits (top 13 bits: `#b1010101010101xxx`) so the "header/trailer length mismatch" is an artifact of the length test coming after the trailer password test and it only recording the last error it detects.

I'm suspicious about the two cells both with array block passwords at the beginning of the block that faulted, and the second word - `0xaaa90083` if interpreted as an array block header leads to a matching array block trailer and all the blocks after that it seem like well-formed blocks.

nbriggs commented:
Some progress -- fixed up "makefreearrayblock()" in Maiko to set the freed memory to 0xBAADF00D, and then ran the STRESSDISPLAY code, with all the checking on.  Died in CHECKARRAYS with tag BEFOREDISPLAY, where the block it was checking on looked like:

```
< B 0x4703aa
This array block: 0x4703aa.  Previous: 0x452390.  Next: 0x48e3c4.
        password: 0x1755     gctype: 0x2   in use: 1
  Header  Length: 61453 cells.

 Trailer  Length: 0 cells.

     0x109733704 (0x  470382) baadf00d
     0x109733708 (0x  470384) baadf00d
     0x10973370c (0x  470386) baadf00d
     0x109733710 (0x  470388) baadf00d
     0x109733714 (0x  47038a) baadf00d
     0x109733718 (0x  47038c) baadf00d
     0x10973371c (0x  47038e) baadf00d
     0x109733720 (0x  470390) baadf00d
     0x109733724 (0x  470392) baadf00d
     0x109733728 (0x  470394) baadf00d
     0x10973372c (0x  470396) baadf00d
     0x109733730 (0x  470398) baadf00d
     0x109733734 (0x  47039a) baadf00d
     0x109733738 (0x  47039c) baadf00d
     0x10973373c (0x  47039e) baadf00d
     0x109733740 (0x  4703a0) baadf00d
     0x109733744 (0x  4703a2) baadf00d
     0x109733748 (0x  4703a4) baadf00d
     0x10973374c (0x  4703a6) baadf00d
     0x109733750 (0x  4703a8) baadf00d
     0x109733754 (0x  4703aa) baadf00d <- array header
     0x109733758 (0x  4703ac) baadf00d
     0x10973375c (0x  4703ae) baadf00d
     0x109733760 (0x  4703b0) baadf00d
     0x109733764 (0x  4703b2) baadf00d
     0x109733768 (0x  4703b4) baadf00d
     0x10973376c (0x  4703b6) baadf00d
     0x109733770 (0x  4703b8) baadf00d
     0x109733774 (0x  4703ba) baadf00d
     0x109733778 (0x  4703bc) baadf00d
     0x10973377c (0x  4703be) baadf00d
     0x109733780 (0x  4703c0) baadf00d
     0x109733784 (0x  4703c2) baadf00d
     0x109733788 (0x  4703c4) baadf00d
     0x10973378c (0x  4703c6) baadf00d
     0x109733790 (0x  4703c8) baadf00d
     0x109733794 (0x  4703ca) baadf00d
     0x109733798 (0x  4703cc) baadf00d
     0x10973379c (0x  4703ce) baadf00d
     0x1097337a0 (0x  4703d0) baadf00d
. . .
     0x10976f734 (0x  48e39a)        0
     0x10976f738 (0x  48e39c)        0
     0x10976f73c (0x  48e39e)        0
     0x10976f740 (0x  48e3a0)        0
     0x10976f744 (0x  48e3a2)        0
     0x10976f748 (0x  48e3a4)        0
     0x10976f74c (0x  48e3a6)        0
     0x10976f750 (0x  48e3a8)        0
     0x10976f754 (0x  48e3aa)        0
     0x10976f758 (0x  48e3ac)        0
     0x10976f75c (0x  48e3ae)        0
     0x10976f760 (0x  48e3b0)        0
     0x10976f764 (0x  48e3b2)        0
     0x10976f768 (0x  48e3b4)        0
     0x10976f76c (0x  48e3b6)        0
     0x10976f770 (0x  48e3b8)        0
     0x10976f774 (0x  48e3ba)        0
     0x10976f778 (0x  48e3bc)        0
     0x10976f77c (0x  48e3be)        0
     0x10976f780 (0x  48e3c0)        0
     0x10976f784 (0x  48e3c2)        0 <- array trailer
     0x10976f788 (0x  48e3c4)        0
     0x10976f78c (0x  48e3c6)        0
     0x10976f790 (0x  48e3c8)        0
```
so somebody had a pointer to block 0x4703aa, but this was smashed by being merged into another block.

The history looks like (cblock: check array block; other self-explanatory.  reclaim ptr is the *data* address, so 2 lisp words after the actual block)
```
cblock@0x46fc26
reclaim: ptr 0x46fc28
makefreearrayblock@0x46fc26
cblock@0x46fc26
cblock@0x46fc26
cblock@0x4703aa
arrayblockmerger 0x46fc26 (962 cells) with following 0x4703aa (497 cells)  -- there's our block getting smashed.
makefreearrayblock@0x46fc26
cblock@0x46fc26
cblock@0x472078
reclaim: ptr 0x47207a
makefreearrayblock@0x472078
cblock@0x471fe0
arrayblockmerger 0x471fe0 (76 cells) with following 0x472078 (516 cells)
makefreearrayblock@0x471fe0
cblock@0x471fe0
cblock@0x471fe0
cblock@0x472480
cblock@0x472a76
reclaim: ptr 0x472a78
makefreearrayblock@0x472a76
cblock@0x47299a
arrayblockmerger 0x47299a (110 cells) with following 0x472a76 (529 cells)
makefreearrayblock@0x47299a
cblock@0x47299a
cblock@0x47299a
cblock@0x472e98
```

masinter reviewed (approved):
I will approve this because we said we would at the 6/40/36 meeting.
We should still review to make sure that the changes are documented and can be found in the change notices in the release message.

I've run HCFILES on the repo successfully with no crash.  
I think Nick wanted to verify this with the maiko changes before merging.

nbriggs commented:
I have not found any smashed array blocks running with checking turned on in Maiko.

masinter commented:
needs a better summary of changes
* Tedit split windows, color, internals
* Rename CHARNAME to be CHARCODE.ENCODE
Parallel to CHARCODE.DECODE.  Definition will later be moved to CMLCHARACTER
* Don't bring color if black
* Change loadup order to eliminate undefined in fuller database
* Pass NONCHARIDENTITY in recursive calls to CHARCODE.ENCODE
* Make the split region really go away by default
* Now with an explicit test for -1
Maybe that will really turn off the splitting region.
* Backquote goof up in debug output
* Fix some tab menu issues, plus a few other things
* Display cleanup for a change on the top line of a pane
* Fix break reported by @pamoroso
* Add back \TEDIT.ACTIVE.WINDOWP
Only called by TEDIT-PROCESS-KILLER (not in MS database), was out of date
* Take down secondary selections
* Eliminate error on quitting
* Cleaned up window closing, DINFO graph seems to work again
* Update TEDIT-MENU.LCOM
* TEDIT-WINDOW--better behavior when closing fails
* NWAY buttons default to unsorted--new items go at the end
Otherwise meta-3 for the 3rd font might pick a different one
* Don't flash caret after closing
* Fix RPLACD
* USER.CM can be specified as an opening property
* Unsplit panes before using an old window for a new file
* Tedit split windows, color, internals
* Rename CHARNAME to be CHARCODE.ENCODE
Parallel to CHARCODE.DECODE.  Definition will later be moved to CMLCHARACTER
* Don't bring color if black
* Change loadup order to eliminate undefined in fuller database
* Pass NONCHARIDENTITY in recursive calls to CHARCODE.ENCODE
* Make the split region really go away by default
* Now with an explicit test for -1
Maybe that will really turn off the splitting region.
* Backquote goof up in debug output
* Fix some tab menu issues, plus a few other things
* Display cleanup for a change on the top line of a pane
* Fix break reported by @pamoroso
* Add back \TEDIT.ACTIVE.WINDOWP
Only called by TEDIT-PROCESS-KILLER (not in MS database), was out of date
* Take down secondary selections
* Eliminate error on quitting
* Cleaned up window closing, DINFO graph seems to work again
* Update TEDIT-MENU.LCOM
* TEDIT-WINDOW--better behavior when closing fails
* NWAY buttons default to unsorted--new items go at the end
Otherwise meta-3 for the 3rd font might pick a different one
* Don't flash caret after closing
* Fix RPLACD
* USER.CM can be specified as an opening property
* Unsplit panes before using an old window for a new file
* Fixes for #2173 #2172 #2171 plus recompile CHARSLOT uses
I was hoping that CHARSLOT/THISLINE was the source of crashes, but I'm still seeing them
* Delete FOO.TEDIT
* Delete TEDIT-FIXFILES
Currently just some hacks that might help in fixing some damaged files.  But not really secure
* typos
* Recompile TEDIT-FILE, to be sure
* Validating thisline, plus a glitch or two
* Updated TEDIT-STRESS for gc/array testing
Checks array space often, preceded by RECLAIM if CHECKARRAYS is set to RECLAIM
* TEDIT-STRESS:  STRESS-INSERT now doesn't display
STRESS-INSERTWINDOW has the previous code
* TEDIT-STRESS:  a little cleanup
STRESSINSERT/STRESSDELETE don't' bring up blank window, added STRESSDELETEWINDOW
* Make sure the stress stream is closed
* TEDIT-STRESS Saves RAND state
* Recompiled TEDIT
* TEDIT-STRESS: defaults CHECKARRAYS NIL, NSYSOUTS 0, ARRAYBLOCKCHECKING T
* EQUATIONS;  Image object shouldn't require a window
*
